[PATCH] powerHAL: Adds functionality to enable/disable double-tap wake gesture.

Adds support for set_feature() function to enable or disable the features.
Double-tap wake gesture is the first feature that has been added which
uses this framework.
Set overlay to have double tap to wake support

diff --git a/modules/power/Android.mk b/modules/power/Android.mk
index 04895aa..674f10d 100644
--- a/modules/power/Android.mk
+++ b/modules/power/Android.mk
@@ -18,8 +18,8 @@
 
 LOCAL_MODULE := power.default
 LOCAL_MODULE_RELATIVE_PATH := hw
-LOCAL_SRC_FILES := power.c
-LOCAL_SHARED_LIBRARIES := liblog
+LOCAL_SRC_FILES := power.c feature.c
+LOCAL_SHARED_LIBRARIES := liblog libcutils
 LOCAL_MODULE_TAGS := optional
 
 include $(BUILD_SHARED_LIBRARY)
diff --git a/modules/power/feature.c b/modules/power/feature.c
new file mode 100644
index 0000000..32a9a91
--- /dev/null
+++ b/modules/power/feature.c
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2015 The CyanogenMod Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <errno.h>
+
+#include <hardware/hardware.h>
+#include <hardware/power.h>
+
+#define LOG_TAG "Legacy PowerHAL"
+#include <utils/Log.h>
+
+#define WAKE_GESTURE_PATH "/sys/module/doubletap2wake/parameters/enable"
+
+static void sysfs_write(char *path, char *s)
+{
+    char buf[80];
+    int len;
+    int fd = open(path, O_WRONLY);
+
+    if (fd < 0) {
+        strerror_r(errno, buf, sizeof(buf));
+        ALOGE("Error opening %s: %s\n", path, buf);
+        return;
+    }
+
+    len = write(fd, s, strlen(s));
+    if (len < 0) {
+        strerror_r(errno, buf, sizeof(buf));
+        ALOGE("Error writing to %s: %s\n", path, buf);
+    }
+
+    close(fd);
+}
+
+void set_feature(struct power_module *module, feature_t feature, int state)
+{
+    switch (feature) {
+    case POWER_FEATURE_DOUBLE_TAP_TO_WAKE:
+        sysfs_write(WAKE_GESTURE_PATH, state ? "1" : "0");
+        break;
+    default:
+        ALOGW("Error setting the feature, it doesn't exist %d\n", feature);
+        break;
+    }
+}
diff --git a/modules/power/feature.h b/modules/power/feature.h
new file mode 100644
index 0000000..0082cc7
--- /dev/null
+++ b/modules/power/feature.h
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2015 The CyanogenMod Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+void set_feature(struct power_module *module, feature_t feature, int state);
diff --git a/modules/power/power.c b/modules/power/power.c
index 308ea0c..5938a67 100644
--- a/modules/power/power.c
+++ b/modules/power/power.c
@@ -24,6 +24,7 @@
 
 #include <hardware/hardware.h>
 #include <hardware/power.h>
+#include "feature.h"
 
 static void power_init(struct power_module *module)
 {
@@ -59,4 +60,5 @@
     .init = power_init,
     .setInteractive = power_set_interactive,
     .powerHint = power_hint,
+    .setFeature = set_feature,
 };
